server/src/main/java/io/crate/execution/engine/sort/Sort.java

```diff
@@ -44,26 +52,42 @@ public static <T> CompletableFuture<List<T>> parallelSort(List<T> list,
         int itemsPerThread = list.size() / Math.max(1, numAvailableThreads);
         if (numAvailableThreads <= 1 || itemsPerThread < minItemsPerThread) {
             list.sort(comparator);
+            allocateBytes.accept(sortOverheadBytes(list.size()));
             return completedFuture(list);
         }
         List<List<T>> partitions = Lists.partition(list, itemsPerThread);
-        ArrayList<CompletableFuture<List<T>>> futures = new ArrayList<>(partitions.size());
+        allocateBytes.accept(RamUsageEstimator.sizeOfObject(partitions)); // Accounts for nested list
+        CompletableFuture<List<T>>[] futures = new CompletableFuture[partitions.size()];
+        allocateBytes.accept(RamUsageEstimator.shallowSizeOf(futures));
+        int index = 0;
         for (List<T> partition : partitions) {
-            futures.add(supplyAsync(() -> {
+            futures[index] = supplyAsync(() -> {
                 partition.sort(comparator);
+                allocateBytes.accept(sortOverheadBytes(partition.size()));
                 return partition;
-            }, executor));
+            }, executor);
+            index++;
         }
-        return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
-            .thenApply(aVoid -> {
+        return CompletableFuture.allOf(futures)
+            .thenApply(_ -> {
 
                 ArrayList<T> result = new ArrayList<>(list.size());
-                ArrayList<List<T>> parts = new ArrayList<>(futures.size());
-                for (int i = 0; i < futures.size(); i++) {
-                    parts.add(futures.get(i).join());
+                ArrayList<List<T>> parts = new ArrayList<>(futures.length);
+                for (int i = 0; i < futures.length; i++) {
+                    parts.add(futures[i].join());
                 }
                 Iterables.addAll(result, Iterables.mergeSorted(parts, comparator));
+                allocateBytes.accept(RamUsageEstimator.sizeOfObject(parts));
+                allocateBytes.accept(RamUsageEstimator.sizeOfObject(result));
```

Comment:
I saw same Integer's in parts/result but we need to account for pointers/shallowSize anyway. 
`sizeOfObject` is used because it can also handle nested list.

Reply:
Aren't the array elements already accounted for and only shallow size of the new arrays would need to be accounted?